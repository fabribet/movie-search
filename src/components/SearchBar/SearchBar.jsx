import React from 'react'
import PropTypes from 'prop-types'
import styles from './styles.module.scss'

const THROTTLING_TIME = process.env.REACT_APP_THROTTLING_TIME_MS

/**
 * SearchBar - React component.
 * Renders a bar with a search input.
 *
 * Properties
 * - handleSearch {func} - The handler for the search action. {Required}
 * - clearSearch {func}  - Handler for the search input clearence. 
 */
export default class SearchBar extends React.Component {
  constructor (props) {
    super(props)
    this.state = {
      query: ''
    }
    this.throttling = null

    this.search = this.search.bind(this)
  }
  
  static get propTypes() {
    return {
      search: PropTypes.func.isRequired,
      clearSearch: PropTypes.func.isRequired
    }
  }

  /**
   * Validates that the query is not empty and calls the handler.
   */
  search() {
    if (this.state.query) {
      this.props.search(this.state.query)
    } else this.props.clearSearch()
  }

  /**
   * Creates a timeout to trigger the search based on the configured throttling time to avoid innecesary
   * calls to the API.
   */
  throttle() {
    const self = this
    if (this.throttling) clearInterval(this.throttling)
    this.throttling = setTimeout (() => {
      self.search()
    }, THROTTLING_TIME)
  }
  
  render() {
    return (
      <div className={styles.searchBar}>
        <form className={styles.inputContainer} >
          <div className={styles.searchIconContainer}>
            <i className="material-icons"></i>
          </div>
          <input
            value={this.state.query}
            onChange={event => {
              this.setState({ query: event.target.value }, () => {
                this.throttle()
              })

            }}
            onKeyDown={event => {
              switch (event.key) {
                case 'Escape':
                  this.setState({ query: '' }, () => {
                    this.throttle()
                  })
                  break;
                default:
                    break
              }
            }}
            className={styles.searchBox}
            type="text"
            placeholder="Search..."
          />
        </form>
      </div>
    )
  }
}