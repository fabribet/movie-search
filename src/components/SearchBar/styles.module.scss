@import '../../scss/colors.scss';

.searchBar {
  background-color: white;
  padding: 5px;
  text-align: center;
  margin-bottom: 20px;
  position: sticky;
  top: 0;
}

.inputContainer {
  display: flex;
  width: 90%;
  max-width: 1000px;
  height: 20px;
  margin: 0 auto;

  .searchIconContainer {
    background-image: url('./imgs/search-solid.svg');
    background-repeat: no-repeat;
    background-size: 15px 15px;
    background-position: center;
    width: 30px;
    background-color: $white;
  }

  .searchBox {
    border: 0;
    flex-grow: 1;
    border-bottom: 1px solid gray;
  
    &:focus {
      border-color: rgb(36, 168, 212);
    }
  }
}

.searchBox:focus {
  outline: 0;
}