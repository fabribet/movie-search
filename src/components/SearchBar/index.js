import { connect } from 'react-redux'
import { actions } from '../../actions/TMDB'
import SearchBar from './SearchBar'

const mapDispatchToProps = dispatch => ({
  /**
   * Dispatches the SearchMovie action.
   * @param {string} query - The search query string.
   */
  search (query) {
    dispatch(actions.SearchMovies(query))
  },
  /**
   * Dispatches the ClearSearch action.
   */
  clearSearch () {
    dispatch(actions.ClearSearch())
  }
})

export default connect(null, mapDispatchToProps)(SearchBar)
