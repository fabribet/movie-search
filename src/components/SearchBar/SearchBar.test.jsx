import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import SearchBar from './SearchBar'

test('renders without crashing', () => {
  const div = document.createElement('div')
  ReactDOM.render(
    <SearchBar search={() => {}} clearSearch={() => {}} />,
    div
  )
})

it('Renders favorite box correctly', () => {
  const tree = renderer
    .create(<SearchBar search={() => {}} clearSearch={() => {}} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

describe('Search input behavior', () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  it('Waits 0.3 seconds before triggering search', () => {
    // Render a checkbox with label in the document
    const shallowSB = shallow(<SearchBar search={() => {}} clearSearch={() => {}} />)
    const input = shallowSB.find('input#search-bar')
    input.simulate('change', { target: { value: 'asd' } })

    expect(setTimeout).toHaveBeenCalledTimes(1)
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), '300')
  })

  it('Search function is called when input has changed', () => {
    const VALUE = 'Search function'
    const searchSpy = jest.fn()

    // Render a SearchBar
    const shallowSB = shallow(<SearchBar search={searchSpy} clearSearch={() => {}} />)

    const input = shallowSB.find('input#search-bar')
    input.simulate('change', { target: { value: VALUE } })
    jest.runAllTimers()

    expect(searchSpy).toHaveBeenCalled()
    expect(searchSpy).toHaveBeenLastCalledWith(VALUE)
    expect(searchSpy).toHaveBeenCalledTimes(1)
  })

  it('Clear function is called when input has been cleared', () => {
    const VALUE = 'clear function'
    const searchSpy = jest.fn()
    const clearSearchSpy = jest.fn()
    // Render a SearchBar
    const shallowSB = shallow(<SearchBar search={searchSpy} clearSearch={clearSearchSpy} />)
    const input = shallowSB.find('input#search-bar')

    input.simulate('change', { target: { value: VALUE } })
    jest.runAllTimers()

    input.simulate('change', { target: { value: '' } })
    jest.runAllTimers()
    expect(clearSearchSpy).toHaveBeenCalled()
    expect(clearSearchSpy).toHaveBeenCalledTimes(1)
  })
})
