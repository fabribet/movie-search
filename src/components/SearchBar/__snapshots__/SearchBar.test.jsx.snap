// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders favorite box correctly 1`] = `
<div
  className="searchBar"
>
  <form
    className="inputContainer"
  >
    <div
      className="searchIconContainer"
    >
      <i
        className="material-icons"
      />
    </div>
    <input
      className="searchBox"
      id="search-bar"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Search..."
      type="text"
      value=""
    />
  </form>
</div>
`;
